Sources/Widgets/Widgets3D/ShapeWidget/Constants.js

```diff
@@ -20,4 +22,72 @@ export const ShapeBehavior = {
   },
 };
 
+export const HorizontalTextPosition = {
+  OUTSIDE_LEFT: 'OUSIDE_LEFT',
+  INSIDE_LEFT: 'INSIDE_LEFT',
+  OUTSIDE_RIGHT: 'OUSIDE_RIGHT',
+  INSIDE_RIGHT: 'INSIDE_RIGHT',
+  MIDDLE: 'MIDDLE',
+};
+
+export const VerticalTextPosition = {
+  OUTSIDE_TOP: 'OUSIDE_TOP',
+  INSIDE_TOP: 'INSIDE_TOP',
+  OUTSIDE_BOTTOM: 'OUSIDE_BOTTOM',
+  INSIDE_BOTTOM: 'INSIDE_BOTTOM',
+  MIDDLE: 'MIDDLE',
+};
+
+export function computeTextPosition(
+  bounds,
+  horizontalPosition,
+  verticalPosition,
+  textWidth,
+  textHeight
+) {
+  let x = 0;
+  switch (horizontalPosition) {
+    case HorizontalTextPosition.OUTSIDE_LEFT:
+      x = bounds[0] - textWidth;
+      break;
+    case HorizontalTextPosition.INSIDE_LEFT:
+      x = bounds[0];
+      break;
+    case HorizontalTextPosition.MIDDLE:
+      x = 0.5 * (bounds[0] + bounds[1] - textWidth);
+      break;
+    case HorizontalTextPosition.INSIDE_RIGHT:
+      x = bounds[1] - textWidth;
+      break;
+    case HorizontalTextPosition.OUTSIDE_RIGHT:
+      x = bounds[1];
+      break;
+    default:
+      break;
+  }
+
+  let y = 0;
+  switch (verticalPosition) {
+    case VerticalTextPosition.OUTSIDE_TOP:
+      y = bounds[3] + textHeight;
+      break;
+    case VerticalTextPosition.INSIDE_TOP:
+      y = bounds[3];
+      break;
+    case VerticalTextPosition.MIDDLE:
+      y = 0.5 * (bounds[2] + bounds[3] + textWidth);
+      break;
+    case VerticalTextPosition.INSIDE_BOTTOM:
+      y = bounds[2] + textHeight;
+      break;
+    case VerticalTextPosition.OUTSIDE_BOTTOM:
+      y = bounds[2];
+      break;
+    default:
+      break;
+  }
+
+  return [x, y, 0];
```

Comment:
bounds[4] for z ?

Reply:
There is no depth in screen coordinates so we do not bother to compute anything.